### PR TITLE
build: Upgrade all Nix deps, Nodejs, Yarn, TypeScript, switch to Yarn PnP and package all beacon-light-client/plonky2 services

### PR DESCRIPTION
**Pull request by PetarKirov**
* [af33c3fa] refactor(beacon-light-client): Delete unfinished typescript prototype - by @PetarKirov
* [07a72137] build(yarn): Upgrade TypeScript to 5.2.2 - by @PetarKirov
* [ba2ebbc8] fix(ts-jest) Upgrade packages and fix issue jest.config - by @PetarKirov
* [7b34134a] build(yarn): Upgrade to v4.1.1 - by @MartinNikov
* [90a00b4c] build(yarn): Change `nonodeLinker` to `pnp` - by @PetarKirov
* [6c7e330f] config(git): Un-ignore .vscode/{extensions,settings}.json - by @PetarKirov
* [aabd74d8] build(yarn|vscode): Enable yarn vscode sdk - by @PetarKirov
* [626e9602] build(nix/yarn): Add `yarn-plugin-nixify` plugin - by @MartinNikov
* [7cb0ff23] improve(balance_verification_circuit): Optimize, change maximum level to 37, add error messages - by @monyarm
* [e28a9229] build(flake.nix): Add `craneLib-nightly` - by @MartinNikov
* [1874c9cf] feat(nix/pkgs): Add `commitment_mapper_builder` - by @MartinNikov
* [d4feb0db] build(flake.nix): Define custom version of `rust-nightly` - by @MartinNikov
* [c51bf829] build(beacon-light-client/plonky2): Update cargo lockfiles - by @MartinNikov
* [8b1e6ddd] config(commitment_mapper_builder): Change name to `circuits_executables` - by @MartinNikov
* [f0b9786b] build(nix): Create `balance-verifier` packages - by @MartinNikov
* [b2956fb2] feat(plonky2): Add `get_balances_inputs` package - by @MartinNikov
* [f991bb4c] build(nix/shell): Change to flake-parts style module and add rust nightly - by @monyarm
* [53fbd6fd] feat(docker-compose): Create compose file for balance verification circuits - by @monyarm
* [682ebd51] partially working docker-compose - by @monyarm
* [fcf067c4] Mock mode - by @monyarm
* [c807365c] Remove unused imports and logs - by @Dimo99
* [2fc38f72] fix mock for changed validators - by @Dimo99
* [7e95729b] improved serialization and fixed some mock bugs in commitment mapper - by @Dimo99
* [e2481928] validators mock problem with loading validators - by @Dimo99
* [d76578ec] move beaconState.bin - by @Dimo99
* [5de6f458] fix bug with changed validators - by @Dimo99
* [f93559ec] Fix some issues - by @monyarm
* [6c094684] Mock changes - by @Dimo99
* [8bdb325b] Reduce get_balances_input output when mock - by @monyarm
* [b346feb0] reduced output when mock - by @monyarm
* [58fec22c] feat(validator-balances): Create disaggregated storage for the proofs in redis - by @Dimo99
* [0fd488b9] build(yarn): Install `colors` - by @PetarKirov
* [11b80fdf] fix(get_changed_validators): Validate mock beacon state binary size - by @PetarKirov
* [8bd2dad8] build(nix): Add `rust-src` and `rust-analyzer` components to the rust-nightly package - by @PetarKirov
* [ab93cd73] build(beacon-light-client/solidity/yarn): Add dotenv - by @PetarKirov
* [1c034bb9] config(prettier): Add more ignored files and dirs - by @PetarKirov
* [185bd6c5] config(yarn/format:check): Exclude `yarn-project.nix` file - by @PetarKirov
* [cf7b02e7] style: Run `yarn format:write` - by @PetarKirov
* [601056ea] ci(gh-actions): rm `magic-nix-cache-action` and update `nix-installer-action` and `cachix-action` - by @PetarKirov
* [e0d76066] fix(beacon-light-client/circom/test): Don't assume that `ts-node` in the $PATH - by @PetarKirov
* [8cb741b2] build(beacon-light-client/solidity): Add missing `@lodestar/types` dep - by @PetarKirov
* [8e98ac01] build(yarn): Upgrade `@chainsafe/ssz` and `@lodestar/types` - by @PetarKirov
* [75972ad5] build(yarn): Add `build:{tsc,sol}` scripts to workspaces - by @PetarKirov
* [cd0058ef] build(typescript): Streamline the configuration - by @PetarKirov
* [93789746] build(yarn): Add missing dependencies - by @PetarKirov
* [32dacaff] fix(typescript): Fix all build errors when running `yarn tsc` - by @PetarKirov
* [1671335b] build(flake.lock): Update all inputs (2024-03-01) - by @PetarKirov
* [1d174f3d] build(nix/shell): Downgrade nim to nim1 - by @PetarKirov
* [068b0764] build(nix): Upgrade Node.js to v21 - by @PetarKirov
* [558de826] build(yarn): Add `ts` script based on `tsx` - by @PetarKirov
* [200554de] fix(beacon-light-client/solidity): Disable `hardhat-gas-reporter` to allow `hardhat test` to pass - by @PetarKirov
* [6f268cb7] build(yarn): Add missing transitive deps - by @PetarKirov
* [6a731d83] build(yarn): Upgrade hardhat - by @PetarKirov
* [7e2a415f] fix(wasm-utils): Add implementation for `emscripten_memcpy_js` - by @PetarKirov
* [ee4dc2f1] refactor(tests/test-nim-to-wasm): Run compile step before each corresponding test - by @PetarKirov
* [7865b2f4] ci(gh-actions): Update `ETHEREUM_MAINNET_RPC` to a working provider - by @PetarKirov
* [628d7ea8] build(nix/circuits_executables): Add necessary build inputs and refactor crane artifacts - by @PetarKirov
* [d36c2ccf] build(nix/{get_balances_input,get_changed_validators}): Use explicit params and reduce repetition - by @PetarKirov
* [3102ca62] build(yarn): Unplug `circomlib` so that it can be physically visible from `circom` - by @PetarKirov
* [53d5167d] build(beacon-light-client/circom): Create `circom` wrapper that passes `-l $CIRCOM_LIB` - by @PetarKirov
* [665b07ba] build(beacon-light-client/circom/yarn): Add `tsx` as dev dep - by @PetarKirov
* [86560263] fix(beacon-light-client/circom/test/run_snarkit2_tests): Erradicate reliance on CWD - by @PetarKirov

### CODE REVIEW

**Review by zah — COMMENTED** [review]
Are the binary files and the various *copy* and *old.nix* files in this PR intended for committing?

**Review by EmilIvanichkovv — APPROVED** [review]
Finally we can 🛳️ it!

File: beacon-light-client/circom/circuits/expand_message.circom [hunk 1]
```diff
@@ -1,7 +1,7 @@
 pragma circom 2.1.5;
 
-include "../../../node_modules/circomlib/circuits/sha256/sha256.circom";
-include "../../../node_modules/circomlib/circuits/bitify.circom";
```
Comment (EmilIvanichkovv) [inline beacon-light-client/circom/circuits/expand_message.circom:4]: 🤢🤢🤢🤢

Nice catch!

File: libs/nix/common-shell-pkgs.nix [hunk 2]
```diff
@@ -36,7 +36,12 @@ in
     # 15% size reduction of the generated .wasm files.
     binaryen
 
-    metacraft-labs.circom
+    (
+      pkgs.writeShellScriptBin "circom" ''
+        ${metacraft-labs.circom}/bin/circom -l "$CIRCOM_LIB" "$@"
```
Comment (EmilIvanichkovv) [inline libs/nix/common-shell-pkgs.nix:41]: Miracle 👼👼👼

**Review by EmilIvanichkovv — APPROVED** [review]

**Review by Xearty — APPROVED** [review]

**Comment by PetarKirov** [comment]
Thank you @EmilIvanichkovv for all the pair-programming sessions!

**Pull request merged**